File: lib/eagleWingsCinematics.js
// ----------

// EAGLE WINGS CINEMATICS SCRAPER

// - parses data from https://www.eaglewingscinematics.com.sg/movies#openModal
// - extracts the following
    // img src
    // details-title
    // details-rating
    // details-description
    // details-book

// - dependancy installation
    // npm install playwright
    //  npx playwright install

// ----- NODEJS SPECIFIC -----

const { chromium } = require('playwright');

async function fetchAndParseHTML(url) {
    try {
        const browser = await chromium.launch();
        const page = await browser.newPage();
        await page.goto(url, { waitUntil: 'networkidle' });
        let previousHeight;
        let currentHeight = await page.evaluate(() => document.body.scrollHeight);
        while (true) {
            previousHeight = currentHeight;
            await page.evaluate(() => window.scrollTo(0, document.body.scrollHeight));
            await page.waitForTimeout(3000); // wait for content to load
            currentHeight = await page.evaluate(() => document.body.scrollHeight);
            if (currentHeight === previousHeight) break;
        }
        const content = await page.content();
        await browser.close();
        return content;
    } catch (error) {
        console.error('Error fetching HTML:', error);
    }
}

// ----- BROWSER SPECIFIC -----
    // archived

// async function fetchAndParseHTML() {
//     function scrollToBottom() {
//         window.scrollTo(0, document.body.scrollHeight);
//     }
//     function getHTML() {
//         return document.documentElement.outerHTML;
//     }
//     function sleep(ms) {
//         return new Promise(resolve => setTimeout(resolve, ms));
//     }
//     async function waitForPageLoad() {
//         let lastScrollHeight = document.body.scrollHeight;
//         let currentScrollHeight = lastScrollHeight;
//         let stableScroll = false;
//         while (!stableScroll) {
//             scrollToBottom();
//             await sleep(3000); 
//             currentScrollHeight = document.body.scrollHeight;
//             if (currentScrollHeight === lastScrollHeight) {
//                 stableScroll = true; 
//             } else {
//                 lastScrollHeight = currentScrollHeight;
//             }
//         }
//     }
//     await waitForPageLoad();
//     const htmlContent = getHTML();
//     console.log('Page HTML:', htmlContent);
//     return htmlContent;
// }

async function extractFilmsEwc(url) {
    const htmlText = await fetchAndParseHTML(url);
    if (htmlText) {
        const { JSDOM } = require('jsdom');
        const dom = new JSDOM(htmlText);
        const doc = dom.window.document;

        // -- debugging -- 

        // console.log(doc.title);
        // console.log(doc.body.innerHTML)

        const eagleWingsCinematicsFilms = doc.querySelectorAll('.movie-list > li');
        let numMovies = 0;

        const eagleWingsCinematicsFilmsObject = [];

        if (eagleWingsCinematicsFilms) {

            eagleWingsCinematicsFilms.forEach(ewcFilm => {

                const imageSrc = ewcFilm.querySelector('img')?.getAttribute('src') || 'N/A';
                const title = ewcFilm.querySelector('.details-title');
                const titleText = title ? title.textContent.trim() : 'N/A';
                const rating = ewcFilm.querySelector('[class^="details-rating"]');
                const ratingText = rating ? rating.textContent.trim() : 'N/A';
                const description = ewcFilm.querySelector('.details-description');
                const descriptionText = description ? description.textContent.trim() : 'N/A';
                const bookTicketsUrl = ewcFilm.querySelector('.details-book')?.getAttribute('href') || 'N/A';
                const bookTicketsUrlText = bookTicketsUrl.trim() || 'N/A';

                // console.log(`
                //         -----
                //         Image src: ${imageSrc}
                //         Title: ${titleText}
                //         Description: ${descriptionText}
                //         Rating : ${ratingText}
                //         Book tickets Url: ${bookTicketsUrlText}
                //         -----
                //     `);

                eagleWingsCinematicsFilmsObject.push({
                    title: titleText,
                    tags: null,
                    description: descriptionText,
                    duration: null,
                    rating: ratingText,
                    bookTicketsUrl: bookTicketsUrlText,
                    posterSrc: imageSrc
                });

                numMovies++;

            })

            console.log(`number of ewc movies => ${numMovies}`);

            return eagleWingsCinematicsFilmsObject;

        } else {
            console.log('No eaglewings cinematics films found');
        }
    } else {
        console.log('HTML content not found');
    }
};

module.exports = { extractFilmsEwc };

// ----- EXECUTION CODE -----
    // - for testing
    // - actual usage to be called from /src/main.js

// extractFilmsEwc("https://www.eaglewingscinematics.com.sg/movies#openModal").then((ewcFilmsObject) => {
//     if (ewcFilmsObject) {
//         console.log(JSON.stringify(ewcFilmsObject, null, 2));
//     } else {
//         console.log('no movies data were found');
//     }
// });

File: rub/eagleWingsCinematics.rb
# ----------

# EAGLE WINGS CINEMATICS SCRAPER

# - parses data from https://www.eaglewingscinematics.com.sg/movies#openModal
# - extracts the following
    # img src
    # details-title
    # details-rating
    # details-description
    # details-book

# ----------

require 'selenium-webdriver'
require 'nokogiri'
require 'json' 

def fetch_and_parse_html(url)
  driver = Selenium::WebDriver.for :chrome
  driver.navigate.to url

  previous_height = driver.execute_script('return document.body.scrollHeight')
  loop do
    driver.execute_script('window.scrollTo(0, document.body.scrollHeight)')
    sleep 3 
    current_height = driver.execute_script('return document.body.scrollHeight')
    break if current_height == previous_height
    previous_height = current_height
  end

  content = driver.page_source
  driver.quit

  content
rescue => e
  puts "Error fetching HTML: #{e.message}"
  nil
end

def extract_films(url)
  html_text = fetch_and_parse_html(url)
  return unless html_text

  doc = Nokogiri::HTML(html_text)
  
  eagle_wings_cinematics_films = doc.css('.movie-list > li')
  num_movies = 0

  eagle_wings_cinematics_films_object = []

  eagle_wings_cinematics_films.each do |ewc_film|
    image_src = ewc_film.at_css('img')&.[]('src') || 'N/A'
    title = ewc_film.at_css('.details-title')
    title_text = title ? title.text.strip : 'N/A'
    rating = ewc_film.at_css('[class^="details-rating"]')
    rating_text = rating ? rating.text.strip : 'N/A'
    description = ewc_film.at_css('.details-description')
    description_text = description ? description.text.strip : 'N/A'
    book_tickets_url = ewc_film.at_css('.details-book')&.[]('href') || 'N/A'
    book_tickets_url_text = book_tickets_url.strip || 'N/A'

    eagle_wings_cinematics_films_object << {
      title: title_text,
      tags: nil,
      description: description_text,
      duration: nil,
      rating: rating_text,
      book_tickets_url: book_tickets_url_text,
      poster_src: image_src
    }

    num_movies += 1
  end

  puts "Number of ewc movies => #{num_movies}"

  puts JSON.pretty_generate(eagle_wings_cinematics_films_object)
rescue => e
  puts "Error extracting films: #{e.message}"
  nil
end

# ----- EXECUTION CODE -----
  # For testing

films = extract_films("https://www.eaglewingscinematics.com.sg/movies#openModal")
if films
  puts JSON.pretty_generate(films)
else
  puts 'No movies data were found'
end